rs_lib/src/file_fetcher/mod.rs

```diff
@@ -945,6 +1025,190 @@ impl<TBlobStore: BlobStore, TSys: FileFetcherSys, THttpClient: HttpClient>
+  }
+
+  fn handle_local(
+    &self,
+    url: &Url,
+    options: &FetchLocalOptions,
+  ) -> Result<Option<FileOrRedirect>, FetchLocalError> {
+    self
+      .0
+      .fetch_local(url, options)
+      .map(|maybe_value| maybe_value.map(FileOrRedirect::File))
+  }
+
+  fn handle_data_url(
+    &self,
+    url: &Url,
+  ) -> Result<FileOrRedirect, DataUrlDecodeError> {
+    self.0.fetch_data_url(url).map(FileOrRedirect::File)
+  }
+
+  async fn handle_blob_url(
+    &self,
+    url: &Url,
+  ) -> Result<FileOrRedirect, FetchNoFollowError> {
+    self.0.fetch_blob_url(url).await.map(FileOrRedirect::File)
+  }
+
+  fn handle_fetch_cached_no_follow(
+    &self,
+    url: &Url,
+    maybe_checksum: Option<Checksum<'_>>,
+  ) -> Result<Option<FileOrRedirect>, FetchCachedNoFollowError> {
+    self.0.fetch_cached_no_follow(url, maybe_checksum)
+  }
+
+  async fn handle_fetch_remote_no_follow_no_cache(
+    &self,
+    url: &Url,
+    maybe_accept: Option<&str>,
+    maybe_checksum: Option<Checksum<'_>>,
+    maybe_auth: Option<(header::HeaderName, header::HeaderValue)>,
+  ) -> Result<FileOrRedirect, FetchNoFollowError> {
+    self
+      .0
+      .fetch_remote_no_follow_no_cache(
+        url,
+        maybe_accept,
+        maybe_checksum,
+        maybe_auth,
+      )
+      .await
+  }
+}
+
+struct EnsureCachedFetchStrategy<
+  'a,
+  TBlobStore: BlobStore,
+  TSys: FileFetcherSys,
+  THttpClient: HttpClient,
+>(&'a FileFetcher<TBlobStore, TSys, THttpClient>);
+
+#[async_trait::async_trait(?Send)]
+impl<TBlobStore: BlobStore, TSys: FileFetcherSys, THttpClient: HttpClient>
+  FetchStrategy
+  for EnsureCachedFetchStrategy<'_, TBlobStore, TSys, THttpClient>
+{
+  type ReturnValue = CachedOrRedirect;
+
+  fn handle_memory_file(&self, _file: File) -> CachedOrRedirect {
+    CachedOrRedirect::Cached
+  }
+
+  fn handle_local(
+    &self,
+    url: &Url,
+    _options: &FetchLocalOptions,
+  ) -> Result<Option<CachedOrRedirect>, FetchLocalError> {
+    let path = url_to_file_path(url)?;
+    let maybe_file = self.0.handle_open_file(url, &path)?;
+    Ok(maybe_file.map(|_| CachedOrRedirect::Cached))
+  }
+
+  fn handle_data_url(
+    &self,
+    _url: &Url,
+  ) -> Result<CachedOrRedirect, DataUrlDecodeError> {
+    Ok(CachedOrRedirect::Cached)
+  }
+
+  async fn handle_blob_url(
+    &self,
+    _url: &Url,
+  ) -> Result<CachedOrRedirect, FetchNoFollowError> {
+    Ok(CachedOrRedirect::Cached)
+  }
+
+  fn handle_fetch_cached_no_follow(
+    &self,
+    url: &Url,
+    _maybe_checksum: Option<Checksum<'_>>,
+  ) -> Result<Option<CachedOrRedirect>, FetchCachedNoFollowError> {
+    // We don't take into account the checksum here because we assume
+    // the bytes were verified when initially downloading the data
+    // from the remote server. This is to prevent loading the data into
+    // memory.
+    if self.0.http_cache.contains(url) {
+      Ok(Some(CachedOrRedirect::Cached))
+    } else {
+      Ok(None)
+    }
+  }
```

Comment:
Pay close attention to this method. Is it ok we don't verify the checksum of the cached module here? Otherwise we need to load it into memory when maybe_checksum.is_some()

Reply:
I think it's okay, imo it's not really different (security wise) from the fact that we don't checksum local source files